src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/13.sql

```diff
@@ -0,0 +1,3443 @@
+CREATE TABLE dbo.ResourceChangeType(
+	ResourceChangeTypeId smallint IDENTITY(0,1) NOT NULL,
+	Name nvarchar(50) NOT NULL,
+ CONSTRAINT PK_ResourceChangeType PRIMARY KEY CLUSTERED 
+(
+	ResourceChangeTypeId ASC
+)WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON, OPTIMIZE_FOR_SEQUENTIAL_KEY = OFF) ON [PRIMARY]
+) ON [PRIMARY]
+GO
+
+SET IDENTITY_INSERT [dbo].[ResourceChangeType] ON 
+
+INSERT dbo.ResourceChangeType (ResourceChangeTypeId, Name) VALUES (0, N'Creation')
+INSERT dbo.ResourceChangeType (ResourceChangeTypeId, Name) VALUES (1, N'Update')
+INSERT dbo.ResourceChangeType (ResourceChangeTypeId, Name) VALUES (2, N'Deletion')
+
+SET IDENTITY_INSERT dbo.ResourceChangeType OFF
+GO
+
+
+/*************************************************************
+    Stored procedures for capturing and fetching resource changes
+**************************************************************/
+--
+-- STORED PROCEDURE
+--     CaptureResourceChanges
+--
+-- DESCRIPTION
+--     Inserts resource change data
+--
+-- PARAMETERS
+--     @isDeleted
+--         * Whether this resource marks the resource as deleted.
+--     @version
+--         * The version of the resource being written
+--     @resourceId
+--         * The resource ID
+--     @resourceTypeId
+--         * The ID of the resource type
+--
+-- RETURN VALUE
+--     It does not return a value.
+--
+CREATE PROCEDURE dbo.CaptureResourceChanges
+    @isDeleted bit, 
+    @version int, 
+    @resourceId varchar(64),
+    @resourceTypeId smallint	
+AS
+BEGIN
+    DECLARE @changeType SMALLINT
+    IF (@isDeleted = 1) BEGIN
+        SET @changeType = 2    /* DELETION */
+    END
+    ELSE BEGIN
+        IF (@version = 1) BEGIN
+            SET @changeType = 0 /* CREATION */
+        END
+        ELSE BEGIN
+            SET @changeType = 1 /* UPDATE */
+        END
+    END
+
+    INSERT INTO dbo.ResourceChangeData
+        (ResourceId, ResourceTypeId, ResourceVersion, ResourceChangeTypeId)
+    VALUES
+        (@resourceId, @resourceTypeId, @version, @changeType)
+END
+GO
+
+--
+-- STORED PROCEDURE
+--     FetchResourceChangeData
+--
+-- DESCRIPTION
+--     Returns the number of resource change records from startId
+--
+-- PARAMETERS
+--     @startId
+--         * The start id of resource change records to fetch.
+--     @pageSize
+--         * The page size for fetching resource change records.
+--
+-- RETURN VALUE
+--     Resource change data rows.
+--
+CREATE PROCEDURE dbo.FetchResourceChangeData
+    @startId BIGINT, 
+    @pageSize BIGINT
+AS
+BEGIN
+
+    SET NOCOUNT ON;
+
+    SET TRANSACTION ISOLATION LEVEL REPEATABLE READ;
+
+    SELECT TOP(@pageSize) c.Id
+      ,c.Timestamp
+      ,c.ResourceId
+      ,c.ResourceTypeId
+      ,t.Name AS ResourceTypeName
+      ,c.ResourceVersion
+      ,c.ResourceChangeTypeId
+      FROM dbo.ResourceChangeData c
+      LEFT JOIN dbo.ResourceType t ON c.ResourceTypeId = t.ResourceTypeId
```

Comment:
I found a similar workflow pattern to handle the ResourceType, given that example, maybe we can do similar thing? https://github.com/microsoft/fhir-server/blob/ecf39a2888cb8449b1085867288153f3f8a8b607/src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirModel.cs#L279

Reply:
Technically, yes we can use a similar approach. As we discussed early, calling two separate queries from PulibshEventsService, one for ResourceType select, and the second one for ResourceChangeData select. We can store resource types in the memory using a dictionary. Then, we can dynamically set a resource type name in the C# code.  

To go with this approach, we have to conduct a performance test. So, we can have enough information to compare both approaches. That would be more related to later tasks conducting performance tests and performance improvement.

Reply:
I see. I'm fine with doing it at a later PR. I think the bottle neck overall will be the select query cause it's a recurring query. The query to get resource type into memory should be one time thing in the cold start, I would worry less comparing to left-join.